src/gui/src/layoutViewer.cpp

```diff
@@ -1071,82 +1071,106 @@ void LayoutViewer::drawInstanceShapes(dbTechLayer* layer,
 
 // Draw the instances' names
 void LayoutViewer::drawInstanceNames(QPainter* painter,
-                                     int font_size,
                                      const std::vector<odb::dbInst*>& insts)
 {
   if (!options_->areInstanceNamesVisible()) {
     return;
   }
-  int minimum_height = 5 * minimumViewableResolution();
+
+  const int minimum_size = 10 * minimumViewableResolution();
   const QTransform initial_xfm = painter->transform();
 
   const QColor text_color = options_->instanceNameColor();
   painter->setPen(QPen(text_color, 0));
   painter->setBrush(QBrush(text_color));
 
   const QFont initial_font = painter->font();
+  const QFont text_font = options_->instanceNameFont();
+  const QFontMetricsF font_metrics(text_font);
+
+  // core cell text should be 10% of cell height
+  static const float size_target = 0.1;
+  // text should not fill more than 90% of the instance height or width
+  static const float size_limit = 0.9;
+  static const float rotation_limit = 0.85; // slightly lower to prevent oscillating rotations when zooming
+
+  // limit non-core text to 1/2.0 (50%) of cell height or width
+  static const float non_core_scale_limit = 2.0;
 
-  // aim for 1/4 the width of master, 1/20 the height of master,
-  // but no more than 50x bigger than all other text
-  static const int width_scalar = 4;
-  static const int height_scalar = 20;
-  static const int max_scalar = 50;
-  // text should not fill more than 90% of the instance width
-  static const double width_limit = 0.9;
+  // minimum pixel height for text
+  static const int minimum_text_height = 10;
+
+  const float font_core_scale_height = size_target*pixels_per_dbu_;
+  const float font_core_scale_width = size_limit*pixels_per_dbu_;
 
   painter->setTransform(QTransform());
+  painter->setFont(text_font);
   for (auto inst : insts) {
     dbMaster* master = inst->getMaster();
-    if (master->getHeight() < minimum_height) {
+    int master_height = master->getHeight();
+    int master_width = master->getHeight();
+
+    if (inst->getMaster()->isCore() && master_height < minimum_size) {
+      // if core cell, just check master height
+      continue;
+    } else if (master_width < minimum_size || master_height < minimum_size) {
       continue;
     }
```

Comment:
The code check height twice for core cells.  I would reorder this as:

    if (master_height < minimum_size) {
      continue;
    } else if (inst->getMaster()->isCore() || master_width < minimum_size) {
      // if core cell, just check master height
      continue;
    }